File: DSL/1a.cpp
#include<stdio.h>
#include<vector>
#include<iostream>

using namespace std;
typedef long long ll;
#define REP(i,n) for(int i=0;i<(int)(n);i++)

class UnionFind {
    public:
    std::vector<ll> parent;
    std::vector<ll> sizes;
    UnionFind(ll N) : parent(N), sizes(N, 1) {
        for(ll i=0; i<N; i++) parent[i]=i;
    }
    ll find(ll x){
        if(x==parent[x]) return x;
        return parent[x]=find(parent[x]);
    }
    void unite(ll x, ll y){
        ll x_class = find(x);
        ll y_class = find(y);
        if(x_class == y_class) return;
        if(sizes[x_class] < sizes[y_class])std::swap(x_class, y_class);
        parent[y_class] = x_class;
        sizes[x_class] += sizes[y_class];
        sizes[y_class] = 0;
    }
    bool same(ll x, ll y){
        return find(x)==find(y);
    }
    ll size(ll x){
        return sizes[find(x)];
    }
};


int main(){
    UnionFind uf(10005);
    int n,q, com, x, y;

    cin >> n >> q;
    REP(i,q){
        cin >> com >> x >> y;
        if(com==0){
            uf.unite(x,y);
        }
        else{
            cout << uf.same(x,y) << endl;
        }
        //REP(i,13)cout << uf.parent[i] << " ";
        //cout << endl;
    }
}

File: DSL_python/1a.py
class UnionFind:
    def __init__(self, n:int):
        self.parent = list(range(n))
        self.size = [1] * n

    def find(self, x:int) -> int:
        if self.parent[x] == x:
            return x
        else:
            self.parent[x] = self.find(self.parent[x])
            return self.parent[x]

    def unite(self, x:int, y:int) -> None:
        root_x = self.find(x)
        root_y = self.find(y)
        if root_x == root_y:
            return
        if self.size[root_x] < self.size[root_y]:
            self.parent[root_x] = root_y
            self.size[root_y] += self.size[root_x]
            self.size[root_x] = 0
        else:
            self.parent[root_y] = root_x
            self.size[root_x] += self.size[root_y]
            self.size[root_y] = 0

    def same(self, x:int, y:int) -> bool:
        return self.find(x) == self.find(y)
    
n,q = map(int, input().split())
uf = UnionFind(n)
for _ in range(q):
    com,x,y = map(int, input().split())
    if com == 0:
        uf.unite(x,y)
    elif com == 1:
        if uf.same(x,y):
            print("1")
        else:
            print("0")